Fix agent re-balancing (bsc#1029297)

matelakat opened the pull request:
The original implementation was broken, and did not rebalance the number
of routers between agents. This new implementation provides a
deterministic way to re-balance the routers between agents.

Commit:
Add tests for agent rebalancing

Add a unit test that exercises agent rebalancing. This indicates that
the rebalancing algorithm does not work. Following changes aim to fix
the issue.

aspiers reviewed (changes requested):
At a glance this seems like quite a significant change.  If it's not appropriate to break it into multiple commits then please can you expand the commit message to explain what the change does, so we don't have to reverse engineer it from scratch whilst reviewing.

Also please can you comment on whether #2 is still relevant, and if so, whether this PR obsoletes it?

Commit:
Remove rebalance implementation

The algorithm for rebalancing was wrong as it did not rebalance the
agents properly. The algorithm was only balancing agent pairs, and it
could not balance the landscape within one run, not even after several
runs. Also it was operating on dictionaries and the logic was hard to
follow.

Commit:
Add router rebalancing implementation (bsc#1029297)

Rebalance the number of routers across agents by looking at all the
agents, and moving one router at a time from the most busy to the least
busy agent until the number of routers is balanced across the agents.

matelakat commented:
@aspiers @aplanas @rhafer Change has been reworked, hopefully for the better, could you guys review this?

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -326,6 +326,101 @@ def l3_agent_rebalance(qclient, noop=False, wait_for_router=True):
+    agent_list = live_agent_list(qclient)
+    router_mover = RouterMover(qclient, noop, wait_for_router)
+
+    while agent_list.calculate_maximal_router_count_difference() > 1:
```
This `while` loop makes sense, but contradicts the commit message which seems to say that only a single router will be moved.  Can you fix the commit message?

matelakat replied:
Fixed

aspiers replied:
Much better, thanks!  In fact, I'd now describe this commit as beautifully clear and clean code :-)

aspiers commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -607,6 +607,40 @@ def test_term_can_be_handled_after_context(self):
+    ]
+
+
+def make_neutron_with_distribution(router_distribution):
```
I think `fake_neutron_with_distribution` would be a bit clearer here.

aspiers reviewed (changes requested):
Just one tiny nitpick, but otherwise looks beautiful to me.

matelakat commented:
rebasing...

matelakat commented:
@aspiers @aplanas is this good to go now?

aspiers commented:
@rhafer Please can you review this? I haven't checked but I'm worried it might conflict with #29, and it's earlier in the queue :)

rhafer reviewed (approved):
There is one thing that I am not 100% sure about and that is the handling of agent that have "admin_state_up" set to "False". Agents with that setting can still be hosting active routers. If I understand it correctly those agents will be completely ignored by the new rebalancing code, right? (Which is the right thing to do I think, it just means you will not end up with a completely balanced setup after running the script, it might worth mentioning that in the commit message or a comment)

Other than that I think this is really nice and clean work. Thanks a lot!

rhafer commented:
> Please can you review this?

Done

> I haven't checked but I'm worried it might conflict with #29, and it's earlier in the queue :)

AFAICS it's doesn't conflict.

aspiers commented:
> AFAICS it's doesn't conflict.

I checked and it doesn't.
